packages/plugin-defender/src/hardhat/propose-upgrade.ts

```diff
@@ -0,0 +1,34 @@
+import '@openzeppelin/hardhat-upgrades/dist/type-extensions';
+import { getChainId, ValidationOptions } from '@openzeppelin/upgrades-core';
+import { AdminClient, ProposalResponse } from 'defender-admin-client';
+import type { ContractFactory } from 'ethers';
+import { FormatTypes } from 'ethers/lib/utils';
+import { HardhatRuntimeEnvironment } from 'hardhat/types';
+import { fromChainId } from './network';
+import { getProposalUrl } from './utils';
+
+export type ProposeUpgradeFunction = (
+  proxyAddress: string,
+  ImplFactory: ContractFactory,
+  opts?: ValidationOptions & { title?: string; description?: string },
+) => Promise<ProposalResponse & { url: string }>;
+
+export function makeProposeUpgrade(hre: HardhatRuntimeEnvironment): ProposeUpgradeFunction {
+  return async function proposeUpgrade(proxyAddress, ImplFactory, opts = {}) {
+    if (!hre.config.defender) {
+      throw new Error(`Missing Defender API key and secret in hardhat config`);
+    }
+    const client = new AdminClient(hre.config.defender);
+
+    const chainId = await getChainId(hre.network.provider);
+    const network = fromChainId(chainId);
+    if (!network) {
```

Comment:
```suggestion
    if (network === undefined) {
```

I generally avoid pure `!` because it only checks falsyness and can miss falsy values that are not `undefined` (e.g. empty string). I accept it for objects because all objects are truthy, but when dealing with strings I do the full comparison.